sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java

```diff
@@ -292,7 +292,7 @@ private WriteResult expandTriggered(PCollection<KV<DestinationT, TableRow>> inpu
         .apply(WithKeys.of((Void) null))
         .setCoder(
             KvCoder.of(
-                VoidCoder.of(), KvCoder.of(TableDestinationCoderV2.of(), StringUtf8Coder.of())))
+                VoidCoder.of(), KvCoder.of(TableDestinationCoderV3.of(), StringUtf8Coder.of())))
```

Comment:
I think this will break some Dataflow users who expect to be able to update their pipelines, and can't if these coders change. @chamikaramj what do you think here?

Reply:
I'm bit confused by this comment (and probably missing something). Isn't this similar to the TableDestinationCoder to TableDestinationCoderV2 change we did in following commit to preserve update compatibility ?

https://github.com/apache/beam/commit/b0e03a33cf0c2c573a2d34d88506e19ebb28c934#diff-b2706c94bc268b2bc2b78820ab23b0fe

Reply:
I'm also not sure, I based myself on the commit referenced above. Why doesn't this preserve update compatibility?

Reply:
@reuvenlax in retrospec looking at both the TableDestinationCoder and TableDestinationCoderV2 I see that they are complexly incompatible and could understand requiring the V2 coder. Am I correct that iso adding another Nullable field (clustering) in the TableDestinationCoderV2 will not break the upgradeability? I think the topic of Coder compatibility on upgrades is not a well understood topic.

Reply:
Coder compatibility is unfortunately not well defined. The best assumption today is that any change to coder is incompatible (Beam needs to define a better story around updating coders).

I don't think a Nullable field helps here. The nullable coder will still expect something there, and there will be nothing there.

We need to either:
  1. Decide that this is important enough that we are willing to break update. 
  2 . Make the use of the new coder somehow contingent on the user using clustering.

Reply:
How can we make that decision ?

Also, what did we do previously ? Did we cluster changes that will unavoidably break update compatibility ?

Reply:
@wscheep is approach (2) described by Reuven viable here ? I.e., Can we only use new coder when user use BQ clustering ?

Reply:
For this particular instance, I think we can safely continue to use TableDestinationCoderV2 since none of the subsequent processing steps in this method reference the clustering configuration. The TableDestinations we code here are used to configure a copy job, which does not need to know about clustering.